File: apps/mobile/lib/services/trisafe_api.dart
import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;
import '../models/auth_models.dart';
import '../models/driver_models.dart';
import '../models/driver_profile_update_models.dart';
import '../models/driver_rating_models.dart';
import '../models/driver_violation_models.dart';
import '../models/fare_models.dart';
import '../models/ride_models.dart';
import '../models/terms_models.dart';
import '../models/passenger_safety_models.dart';
import '../models/vehicle_models.dart';

class TriSafeApi {
  final String baseUrl;
  String? _accessToken;

  TriSafeApi({required this.baseUrl});

  Map<String, String> get _headers => {
        'content-type': 'application/json',
        if (_accessToken != null) 'authorization': 'Bearer $_accessToken',
      };

  bool get isAuthenticated => _accessToken != null;

  void logout() {
    _accessToken = null;
  }

  Future<AuthSession> login({
    required String identifier,
    required String password,
    String? expectedRole,
  }) async {
    final session = AuthSession.fromJson(await _postPublic('/auth/login', {
      'identifier': identifier.trim(),
      'password': password,
      if (expectedRole != null) 'expectedRole': expectedRole,
    }));
    _accessToken = session.accessToken;
    return session;
  }

  Future<dynamic> _postPublic(String path, Map<String, dynamic> body) async {
    try {
      final response = await http
          .post(Uri.parse('$baseUrl$path'),
              headers: {'content-type': 'application/json'},
              body: jsonEncode(body))
          .timeout(const Duration(seconds: 10));
      if (response.statusCode >= 400) {
        throw Exception(
            'TriSafe API returned ${response.statusCode}: ${response.body}');
      }
      if (response.body.trim().isEmpty) return null;
      return jsonDecode(response.body);
    } on TimeoutException {
      throw Exception(_connectionMessage());
    } on http.ClientException {
      throw Exception(_connectionMessage());
    }
  }

  Future<dynamic> _get(String path) async {
    return _getUri(Uri.parse('$baseUrl$path'));
  }

  Future<dynamic> _getUri(Uri uri) async {
    try {
      final response = await http
          .get(uri, headers: _headers)
          .timeout(const Duration(seconds: 10));
      if (response.statusCode >= 400) {
        throw Exception(
            'TriSafe API returned ${response.statusCode}: ${response.body}');
      }
      if (response.body.trim().isEmpty) return null;
      return jsonDecode(response.body);
    } on TimeoutException {
      throw Exception(_connectionMessage());
    } on http.ClientException {
      throw Exception(_connectionMessage());
    }
  }

  Future<dynamic> _post(String path, Map<String, dynamic> body) async {
    try {
      final response = await http
          .post(Uri.parse('$baseUrl$path'),
              headers: _headers, body: jsonEncode(body))
          .timeout(const Duration(seconds: 10));
      if (response.statusCode >= 400) {
        throw Exception(
            'TriSafe API returned ${response.statusCode}: ${response.body}');
      }
      return jsonDecode(response.body);
    } on TimeoutException {
      throw Exception(_connectionMessage());
    } on http.ClientException {
      throw Exception(_connectionMessage());
    }
  }

  Future<dynamic> _patch(String path, Map<String, dynamic> body) async {
    try {
      final response = await http
          .patch(Uri.parse('$baseUrl$path'),
              headers: _headers, body: jsonEncode(body))
          .timeout(const Duration(seconds: 10));
      if (response.statusCode >= 400) {
        throw Exception(
            'TriSafe API returned ${response.statusCode}: ${response.body}');
      }
      return jsonDecode(response.body);
    } on TimeoutException {
      throw Exception(_connectionMessage());
    } on http.ClientException {
      throw Exception(_connectionMessage());
    }
  }

  Future<dynamic> _delete(String path) async {
    final response = await http
        .delete(Uri.parse('$baseUrl$path'), headers: _headers)
        .timeout(const Duration(seconds: 10));
    if (response.statusCode >= 400) {
      throw Exception(
          'TriSafe API returned ${response.statusCode}: ${response.body}');
    }
    return response.body.trim().isEmpty ? null : jsonDecode(response.body);
  }

  String _connectionMessage() =>
      'TriSafe API is unavailable at $baseUrl. Start npm run dev:api. Use adb reverse tcp:3000 tcp:3000 only when testing on a wired Android phone.';

  Future<QrVerificationResult> verifyQr(String token) async =>
      QrVerificationResult.fromJson(await _get('/vehicles/verify/$token'));
  Future<PassengerProfile> accountProfile() async =>
      PassengerProfile.fromJson(await _get('/auth/me'));
  Future<PassengerProfile> updatePassengerProfile(
          {String? phone,
          String? email,
          String? avatarData,
          bool updateAvatar = false}) async =>
      PassengerProfile.fromJson(await _patch('/auth/me/profile', {
        if (phone != null) 'phone': phone,
        if (email != null) 'email': email,
        if (updateAvatar) 'avatarData': avatarData
      }));
  Future<DriverProfile> driverProfile() async =>
      DriverProfile.fromJson(await _get('/drivers/me'));
  Future<List<DriverAnnouncement>> driverAnnouncements() async =>
      (await _get('/drivers/me/announcements'))
          .map<DriverAnnouncement>((item) => DriverAnnouncement.fromJson(item))
          .toList();
  Future<List<DriverNotification>> driverNotifications() async =>
      (await _get('/drivers/me/notifications'))
          .map<DriverNotification>((item) => DriverNotification.fromJson(item))
          .toList();
  Future<List<DriverViolationRecord>> driverViolations() async =>
      (await _get('/drivers/me/violations'))
          .map<DriverViolationRecord>(
              (item) => DriverViolationRecord.fromJson(item))
          .toList();
  Future<DriverRatingStatistics> driverRatingStatistics() async =>
      DriverRatingStatistics.fromJson(await _get('/ratings/driver/me'));
  Future<void> markDriverAnnouncementRead(String announcementId) async {
    await _patch('/drivers/me/announcements/$announcementId/read', {});
  }

  Future<void> markDriverNotificationRead(String notificationId) async {
    await _patch('/drivers/me/notifications/$notificationId/read', {});
  }

  Future<void> markAllDriverNotificationsRead() async {
    await _patch('/drivers/me/notifications/read-all', {});
  }

  Future<void> changePassword({
    required String currentPassword,
    required String newPassword,
  }) async {
    await _patch('/auth/me/password', {
      'currentPassword': currentPassword,
      'newPassword': newPassword,
    });
  }

  Future<PublishedTermsDocument?> currentTerms() async {
    final response = await _get('/terms/current');
    if (response == null) return null;
    return PublishedTermsDocument.fromJson(response as Map<String, dynamic>);
  }

  Future<void> updateDriverContact({required String phone}) async {
    await _patch('/drivers/me/contact', {'phone': phone});
  }

  Future<List<DriverLocationOption>> driverMunicipalities() async =>
      (await _get('/drivers/me/locations/bohol/municipalities'))
          .map<DriverLocationOption>(
              (item) => DriverLocationOption.fromJson(item))
          .toList();

  Future<List<DriverLocationOption>> driverBarangays(
          String municipalityCode) async =>
      (await _get(
              '/drivers/me/locations/bohol/municipalities/${Uri.encodeComponent(municipalityCode)}/barangays'))
          .map<DriverLocationOption>(
              (item) => DriverLocationOption.fromJson(item))
          .toList();

  Future<DriverProfile> updateDriverProfile({
    String? phone,
    String? avatarData,
    bool updateAvatar = false,
    DriverAddressUpdate? address,
  }) async {
    final body = <String, dynamic>{
      if (phone != null) 'phone': phone,
      if (updateAvatar) 'avatarData': avatarData,
      if (address != null) 'address': address.toJson(),
    };
    return DriverProfile.fromJson(await _patch('/drivers/me/profile', body));
  }

  Future<List<LocationOption>> locations() async => (await _get('/locations'))
      .map<LocationOption>((item) => LocationOption.fromJson(item))
      .toList();
  Future<FareEstimate> estimateFare(
          {required String vehicleId,
          required String fromLocationId,
          required String toLocationId}) async =>
      FareEstimate.fromJson(await _post('/rides/preview', {
        'vehicleId': vehicleId,
        'fromLocationId': fromLocationId,
        'toLocationId': toLocationId,
      }));
  Future<FareEstimate> estimateDistanceFare({
    required String vehicleType,
    required String passengerType,
    required double originLatitude,
    required double originLongitude,
    required double destinationLatitude,
    required double destinationLongitude,
  }) async =>
      FareEstimate.fromJson(await _post('/distance-fare-estimates', {
        'vehicleType': vehicleType,
        'passengerType': passengerType,
        'originLatitude': originLatitude,
        'originLongitude': originLongitude,
        'destinationLatitude': destinationLatitude,
        'destinationLongitude': destinationLongitude,
      }));
  Future<FareLocationName> fareLocationName({
    required double latitude,
    required double longitude,
  }) async =>
      FareLocationName.fromJson(await _post('/fare-location-names', {
        'latitude': latitude,
        'longitude': longitude,
      }));
  Future<Ride> startRide(
          {required String vehicleId,
          required String fromLocationId,
          required String toLocationId,
          int passengerCount = 1,
          double? startLatitude,
          double? startLongitude}) async =>
      Ride.fromJson(await _post('/rides', {
        'vehicleId': vehicleId,
        'fromLocationId': fromLocationId,
        'toLocationId': toLocationId,
        'passengerCount': passengerCount,
        if (startLatitude != null) 'startLatitude': startLatitude,
        if (startLongitude != null) 'startLongitude': startLongitude,
      }));
  Future<Ride> startMapRide({
    required String vehicleId,
    required String qrToken,
    required double originLatitude,
    required double originLongitude,
    required double destinationLatitude,
    required double destinationLongitude,
    String? originLocationName,
    String? destinationLocationName,
    required String passengerType,
    int passengerCount = 1,
  }) async =>
      Ride.fromJson(await _post('/rides/map', {
        'vehicleId': vehicleId,
        'qrToken': qrToken,
        'originLatitude': originLatitude,
        'originLongitude': originLongitude,
        'destinationLatitude': destinationLatitude,
        'destinationLongitude': destinationLongitude,
        if (originLocationName != null)
          'originLocationName': originLocationName,
        if (destinationLocationName != null)
          'destinationLocationName': destinationLocationName,
        'passengerCount': passengerCount,
        'passengerType': passengerType,
      }));
  Future<Ride> endRide(String rideId,
          {double? endLatitude, double? endLongitude}) async =>
      Ride.fromJson(await _post('/rides/$rideId/end', {
        if (endLatitude != null) 'endLatitude': endLatitude,
        if (endLongitude != null) 'endLongitude': endLongitude,
      }));
  Future<void> updatePresence(
      {required double latitude,
      required double longitude,
      double? accuracy,
      double? heading,
      double? speed}) async {
    await _post('/presence/me', {
      'latitude': latitude,
      'longitude': longitude,
      if (accuracy != null) 'accuracy': accuracy,
      if (heading != null) 'heading': heading,
      if (speed != null) 'speed': speed,
    });
  }

  Future<RideProgress> recordRideLocation(String rideId,
          {required double latitude,
          required double longitude,
          double? accuracy,
          double? heading,
          double? speed}) async =>
      RideProgress.fromJson(await _post('/rides/$rideId/location', {
        'latitude': latitude,
        'longitude': longitude,
        if (accuracy != null) 'accuracy': accuracy,
        if (heading != null) 'heading': heading,
        if (speed != null) 'speed': speed,
      }));
  Future<List<Ride>> rideHistory({DateTime? from, DateTime? to}) async {
    final uri = Uri.parse('$baseUrl/rides').replace(queryParameters: {
      if (from != null) 'from': from.toUtc().toIso8601String(),
      if (to != null) 'to': to.toUtc().toIso8601String(),
    });
    return (await _getUri(uri))
        .map<Ride>((item) => Ride.fromJson(item as Map<String, dynamic>))
        .toList();
  }

  Future<Map<String, dynamic>> shareRide(
    String rideId, {
    double? latitude,
    double? longitude,
  }) async {
    final uri = Uri.parse('$baseUrl/rides/$rideId/share').replace(
      queryParameters: latitude != null && longitude != null
          ? {
              'latitude': latitude.toString(),
              'longitude': longitude.toString(),
            }
          : null,
    );
    return Map<String, dynamic>.from(await _getUri(uri));
  }

  Future<Map<String, dynamic>> draftIncident(String description,
          {String? rideId,
          String? category,
          String? evidenceData,
          String? evidenceName}) async =>
      Map<String, dynamic>.from(await _post('/incidents/draft', {
        'rawDescription': description,
        if (rideId != null) 'rideId': rideId,
        if (category != null) 'category': category,
        if (evidenceData != null) 'evidenceData': evidenceData,
        if (evidenceName != null) 'evidenceName': evidenceName,
      }));
  Future<Map<String, dynamic>> submitIncident(String incidentId,
          {String? finalDescription, String? category}) async =>
      Map<String, dynamic>.from(await _post('/incidents/$incidentId/submit', {
        if (finalDescription != null) 'finalDescription': finalDescription,
        if (category != null) 'category': category
      }));
  Future<List<PassengerIncident>> incidentHistory() async =>
      (await _get('/incidents'))
          .map<PassengerIncident>((item) =>
              PassengerIncident.fromJson(item as Map<String, dynamic>))
          .toList();
  Future<void> createRating(
          {required String rideId,
          required int score,
          String? comment}) async =>
      _post('/ratings', {
        'rideId': rideId,
        'score': score,
        if (comment?.trim().isNotEmpty == true) 'comment': comment!.trim()
      });
  Future<List<TrustedContact>> trustedContacts() async =>
      (await _get('/safety/trusted-contacts'))
          .map<TrustedContact>(
              (item) => TrustedContact.fromJson(item as Map<String, dynamic>))
          .toList();
  Future<TrustedContact> saveTrustedContact(
          {String? id,
          required String fullName,
          required String relationship,
          required String phone,
          bool active = true}) async =>
      TrustedContact.fromJson(id == null
          ? await _post('/safety/trusted-contacts', {
              'fullName': fullName,
              'relationship': relationship,
              'phone': phone,
              'active': active
            })
          : await _patch('/safety/trusted-contacts/$id', {
              'fullName': fullName,
              'relationship': relationship,
              'phone': phone,
              'active': active
            }));
  Future<void> deleteTrustedContact(String id) async =>
      _delete('/safety/trusted-contacts/$id');
  Future<List<dynamic>> emergencyContacts() async =>
      List<dynamic>.from(await _get('/safety/emergency-contacts'));
}

File: apps/mobile/src/services/trisafeApi.ts
import * as SecureStore from "expo-secure-store";
import type {
  AuthSession,
  DriverAnnouncement,
  DriverNotification,
  DriverProfile,
  DriverRatingStatistics,
  DriverViolation,
  EmergencyContact,
  FareEstimate,
  LocationOption,
  PassengerFareType,
  PassengerIncident,
  PassengerProfile,
  QrVerificationResult,
  Ride,
  RideProgress,
  TermsDocument,
  TrustedContact,
} from "../models/trisafe";

const TOKEN_KEY = "trisafe_access_token";
const DEFAULT_API_URL = "http://127.0.0.1:3000/api";

export class TriSafeApi {
  readonly baseUrl: string;
  private accessToken: string | null = null;

  constructor(
    baseUrl = process.env.EXPO_PUBLIC_API_BASE_URL ?? DEFAULT_API_URL,
  ) {
    this.baseUrl = baseUrl.replace(/\/$/, "");
  }

  async restoreSession() {
    this.accessToken = await SecureStore.getItemAsync(TOKEN_KEY);
    if (!this.accessToken) return null;
    try {
      return await this.profile();
    } catch {
      await this.logout();
      return null;
    }
  }

  async login(
    identifier: string,
    password: string,
    expectedRole: "PASSENGER" | "DRIVER",
  ) {
    const session = await this.request<AuthSession>("/auth/login", {
      method: "POST",
      body: { identifier: identifier.trim(), password, expectedRole },
      authenticated: false,
    });
    this.accessToken = session.accessToken;
    await SecureStore.setItemAsync(TOKEN_KEY, session.accessToken);
    return session;
  }

  async logout() {
    this.accessToken = null;
    await SecureStore.deleteItemAsync(TOKEN_KEY);
  }

  profile() {
    return this.request<PassengerProfile>("/auth/me");
  }

  updatePassengerProfile(
    input: Partial<
      Pick<
        PassengerProfile,
        "fullName" | "username" | "phone" | "email" | "avatarData"
      >
    >,
  ) {
    return this.request<PassengerProfile>("/auth/me/profile", {
      method: "PATCH",
      body: input,
    });
  }

  changePassword(currentPassword: string, newPassword: string) {
    return this.request<{ changed: boolean }>("/auth/me/password", {
      method: "PATCH",
      body: { currentPassword, newPassword },
    });
  }

  verifyQr(token: string) {
    return this.request<QrVerificationResult>(
      `/vehicles/verify/${encodeURIComponent(token)}`,
    );
  }

  driverProfile() {
    return this.request<DriverProfile>("/drivers/me");
  }

  driverAnnouncements() {
    return this.request<DriverAnnouncement[]>("/drivers/me/announcements").then(
      (rows) => rows.map(normalizeAnnouncement),
    );
  }

  driverNotifications() {
    return this.request<DriverNotification[]>("/drivers/me/notifications");
  }

  driverViolations() {
    return this.request<DriverViolation[]>("/drivers/me/violations");
  }

  driverRatingStatistics() {
    return this.request<DriverRatingStatistics>("/ratings/driver/me");
  }

  markDriverAnnouncementRead(id: string) {
    return this.request(
      `/drivers/me/announcements/${encodeURIComponent(id)}/read`,
      { method: "PATCH", body: {} },
    );
  }

  markDriverNotificationRead(id: string) {
    return this.request(
      `/drivers/me/notifications/${encodeURIComponent(id)}/read`,
      { method: "PATCH", body: {} },
    );
  }

  markAllDriverNotificationsRead() {
    return this.request("/drivers/me/notifications/read-all", {
      method: "PATCH",
      body: {},
    });
  }

  updateDriverProfile(input: {
    phone?: string;
    avatarData?: string | null;
    address?: Record<string, string>;
  }) {
    return this.request<DriverProfile>("/drivers/me/profile", {
      method: "PATCH",
      body: input,
    });
  }

  driverMunicipalities() {
    return this.request<LocationOption[]>(
      "/drivers/me/locations/bohol/municipalities",
    );
  }

  driverBarangays(municipalityCode: string) {
    return this.request<LocationOption[]>(
      `/drivers/me/locations/bohol/municipalities/${encodeURIComponent(municipalityCode)}/barangays`,
    );
  }

  estimateDistanceFare(input: {
    vehicleType: string;
    passengerType: PassengerFareType;
    originLatitude: number;
    originLongitude: number;
    destinationLatitude: number;
    destinationLongitude: number;
  }) {
    return this.request<FareEstimate>("/distance-fare-estimates", {
      method: "POST",
      body: input,
    });
  }

  fareLocationName(latitude: number, longitude: number) {
    return this.request<{ name: string; context: string }>(
      "/fare-location-names",
      {
        method: "POST",
        body: { latitude, longitude },
      },
    );
  }

  startRide(input: {
    vehicleId: string;
    qrToken: string;
    originLatitude: number;
    originLongitude: number;
    destinationLatitude: number;
    destinationLongitude: number;
    originLocationName?: string;
    destinationLocationName?: string;
    passengerType: PassengerFareType;
    passengerCount: number;
  }) {
    return this.request<Ride>("/rides/map", { method: "POST", body: input });
  }

  endRide(id: string, endLatitude?: number, endLongitude?: number) {
    return this.request<Ride>(`/rides/${encodeURIComponent(id)}/end`, {
      method: "POST",
      body: { endLatitude, endLongitude },
    });
  }

  recordRideLocation(
    id: string,
    input: {
      latitude: number;
      longitude: number;
      accuracy?: number | null;
      heading?: number | null;
      speed?: number | null;
    },
  ) {
    return this.request<RideProgress>(
      `/rides/${encodeURIComponent(id)}/location`,
      { method: "POST", body: input },
    );
  }

  rideHistory(from?: Date, to?: Date) {
    const query = new URLSearchParams();
    if (from) query.set("from", from.toISOString());
    if (to) query.set("to", to.toISOString());
    const suffix = query.toString();
    return this.request<Ride[]>(`/rides${suffix ? `?${suffix}` : ""}`);
  }

  shareRide(id: string, latitude?: number, longitude?: number) {
    const query =
      latitude !== undefined && longitude !== undefined
        ? `?latitude=${latitude}&longitude=${longitude}`
        : "";
    return this.request<Record<string, unknown>>(
      `/rides/${encodeURIComponent(id)}/share${query}`,
    );
  }

  createRating(rideId: string, score: number, comment?: string) {
    return this.request("/ratings", {
      method: "POST",
      body: { rideId, score, comment: comment?.trim() || undefined },
    });
  }

  incidentHistory() {
    return this.request<PassengerIncident[]>("/incidents");
  }

  draftIncident(input: {
    rawDescription: string;
    rideId?: string;
    category?: string;
    evidenceData?: string;
    evidenceName?: string;
  }) {
    return this.request<PassengerIncident>("/incidents/draft", {
      method: "POST",
      body: input,
    });
  }

  updateIncidentDraft(
    id: string,
    input: {
      rawDescription: string;
      category?: string;
      evidenceData?: string;
      evidenceName?: string;
      removeEvidence?: boolean;
    },
  ) {
    return this.request<PassengerIncident>(
      `/incidents/${encodeURIComponent(id)}/draft`,
      { method: "PATCH", body: input },
    );
  }

  submitIncident(id: string, finalDescription: string, category?: string) {
    return this.request<PassengerIncident>(
      `/incidents/${encodeURIComponent(id)}/submit`,
      {
        method: "POST",
        body: { finalDescription, category },
      },
    );
  }

  trustedContacts() {
    return this.request<TrustedContact[]>("/safety/trusted-contacts");
  }

  saveTrustedContact(input: Omit<TrustedContact, "id"> & { id?: string }) {
    const { id, ...body } = input;
    return this.request<TrustedContact>(
      id
        ? `/safety/trusted-contacts/${encodeURIComponent(id)}`
        : "/safety/trusted-contacts",
      {
        method: id ? "PATCH" : "POST",
        body,
      },
    );
  }

  deleteTrustedContact(id: string) {
    return this.request(`/safety/trusted-contacts/${encodeURIComponent(id)}`, {
      method: "DELETE",
    });
  }

  emergencyContacts() {
    return this.request<EmergencyContact[]>("/safety/emergency-contacts");
  }

  currentTerms() {
    return this.request<TermsDocument | null>("/terms/current", {
      authenticated: false,
    });
  }

  private async request<T = unknown>(
    path: string,
    options: RequestOptions = {},
  ) {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10_000);
    try {
      const headers: Record<string, string> = { Accept: "application/json" };
      if (options.body !== undefined)
        headers["Content-Type"] = "application/json";
      if (options.authenticated !== false && this.accessToken)
        headers.Authorization = `Bearer ${this.accessToken}`;
      const response = await fetch(`${this.baseUrl}${path}`, {
        method: options.method ?? "GET",
        headers,
        body:
          options.body === undefined ? undefined : JSON.stringify(options.body),
        signal: controller.signal,
      });
      const text = await response.text();
      const payload: unknown = text ? JSON.parse(text) : null;
      if (!response.ok)
        throw new TriSafeApiError(response.status, apiMessage(payload));
      return payload as T;
    } catch (error) {
      if (error instanceof TriSafeApiError) throw error;
      throw new Error(
        `TriSafe cannot reach ${this.baseUrl}. Check the API and device connection.`,
      );
    } finally {
      clearTimeout(timeout);
    }
  }
}

type RequestOptions = {
  method?: "GET" | "POST" | "PATCH" | "DELETE";
  body?: unknown;
  authenticated?: boolean;
};

export class TriSafeApiError extends Error {
  constructor(
    readonly status: number,
    message: string,
  ) {
    super(message);
  }
}

function apiMessage(payload: unknown) {
  if (!payload || typeof payload !== "object" || !("message" in payload))
    return "TriSafe request failed.";
  const message = (payload as { message: unknown }).message;
  return Array.isArray(message) ? message.join(" ") : String(message);
}

function normalizeAnnouncement(
  row:
    | DriverAnnouncement
    | { announcement: DriverAnnouncement; readAt?: string | null },
) {
  if (!("announcement" in row)) return row;
  return { ...row.announcement, readAt: row.readAt ?? null };
}

export const trisafeApi = new TriSafeApi();
